contracts/FlowIDTableStaking.cdc

```diff
@@ -870,6 +870,79 @@ pub contract FlowIDTableStaking {
             }
         }
 
+        /// Set slot (count) limits for each node type (role).
+        pub fun setSlotLimits(slotLimits: {UInt8: UInt8}) {
+            FlowIDTableStaking.account.load<{UInt8: UInt8}>(from: /storage/flowStakingSlotLimits)
+            FlowIDTableStaking.account.save(slotLimits, to: /storage/flowStakingSlotLimits)
+        }
+
+        /// Enforces slot (count) limits for each node type (role). If more nodes
+        /// of a given type are registered than the limit, they are removed
+        /// randomly until the limits are met. When a node is removed, it's
+        /// committed tokens are transfered to it's unstaked bucket.
+        pub fun slotSelect() {
```

Comment:
When do we expect `slotSelect` to be called in relation to the rest of the staking/epoch flow?

Currently slot selection will remove excess nodes which have sufficient stake and which have `initialWeight == 100`. However, currently all nodes must be a member of the approved list before they are provisionally registered for an epoch (have `initialWeight=100` set). It is unclear to me how we intend for unauthorized Access Nodes to be provisionally registered for an epoch (the step before slot selection).

Reply:
@jordanschalm Yes, you're pointing to an issue.

@Wolog2021 How I expected this to work is as follows:

- During `removedUnapprovedNodes` the contract should track nodes which are transitioning from 0 weight to non-zero weight. This happens on [Line 868](https://github.com/onflow/flow-core-contracts/blob/ea44a5dd9838e44c92119ce470eb724c22772f2e/contracts/FlowIDTableStaking.cdc#L868).

- Then, randomly select from that list of nodes which have transitioned into being available until nodeCounts[type] >= slotLimits[type[.

I believe this whole solution needs to be modified to take this list of possible nodes, and then be called from the end of removeUnapprovedNodes.

Reply:
We can certainly call the function from the end of removeUnapprovedNodes. The reason I didn't do that is that it's doing something fundamentally different i.e. excess otherwise approved nodes are being removed randomly and not because they are unapproved. I didn't want the function name to not be fully descriptive of the operation. Is that a concern here?

Reply:
We can change the name of the function if we want